### ISSUE TITLE
Provide priority mechanism for sorting calls of addons

### ISSUE THREAD

**Issue by dvojtise** [eclipse/gemoc-studio-modeldebugging#111]
given an engine notification, some addons should be called before or after some other addons.

Currently, we have no way to sort them.

Some difficult point, we may not know about the other addons, (so we cannot just create a dependency between addon), the priority may be different on the various calls (ex: high priority on  engine start, low priority on executeStep) 

So I suggest a mechanism based on priorities that are set on each notification.
By default, the priority would be "0", and the higher it is, the earlier it should be run, the lower it is (negative) the later it should be run.

Then before launching a notification, it will sort the addons to be run according to these priorities directives.

**Comment by ebousse** [comment]
This is a real concern indeed! And I fully agree that this must be done for each notification kind.

However I have the feeling that in using integers is rather limited, since it's impossible to give a meaningful absolute value for a priority. What does "10" or "100" mean?

Maybe instead one could explicitly declare dependencies among addons, something like _"this notification must be done before the addon named OmniscientDebugger"_. This would create dependencies among addons, but I have the feeling we would end up doing the same with integers, since one would always look at the priority numbers of other addons when deciding the priority number of its addon.

**Comment by szschaler** [comment]
Would it make sense to add a layer of indirection by defining an (extensible) list of meaningful execution phases and allowing addons to say which execution phase they would like to be in? A little like what Eclipse does with locations for menu items?

**Comment by dvojtise** [comment]
If not using integers, I think that the add-on should also be identified using tags. So tags could be used to find the add-ons that provide a service rather a unique add-on ID.

**Comment by szschaler** [comment]
I wasn't actually thinking about anything that dramatic. Your idea of 'tags' seems very similar to what I was thinking about...

**Comment by dvojtise** [comment]
I have another annoying situation where the addons notifications should be sorted:

It involves 2 **implicit** addons: EngineStatusView and ModelDebugger:

concerning the notifyAboutToExecuteLogicalStep()
The EngineStatusView is in charge of displaying information, but as the ModelDebugger addon is called before, it will lock the execution until the next debug command (step or resume), the call to the EngineStatusView addon in order to refresh it will then occur too late 

![image](https://user-images.githubusercontent.com/661468/72158659-0a202080-33bb-11ea-9bf5-0d81094fbc27.png)
